cmd/bootnode/bootnode.go

```diff
@@ -0,0 +1,201 @@
+		if !config.AutoP2PKey {
+			return errors.New("charon-enr-private-key not found in data dir (run with --auto-p2pkey to auto generate)")
+		}
+
+		log.Info(ctx, "Automatically creating charon-enr-private-key", z.Str("path", p2p.KeyPath(config.DataDir)))
+
+		key, err = p2p.NewSavedPrivKey(config.DataDir)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+
+	// Setup p2p udp discovery.
+	localEnode, db, err := p2p.NewLocalEnode(config.P2PConfig, key)
+	if err != nil {
+		return errors.Wrap(err, "failed to open enode")
+	}
+	defer db.Close()
+
+	udpNode, err := p2p.NewUDPNode(ctx, config.P2PConfig, localEnode, key, nil)
+	if err != nil {
+		return err
+	}
+	defer udpNode.Close()
+
+	// Setup p2p tcp relay (async for snappy startup)
+	var (
+		p2pErr = make(chan error, 1)
+		logP2P = func() {}
+	)
+
+	go func() {
+		if !config.P2PRelay {
+			log.SetLokiLabels(nil)
+			return
+		}
+
+		if config.RelayLogLevel != "" {
+			if err := relaylog.SetLogLevel("relay", config.RelayLogLevel); err != nil {
+				p2pErr <- errors.Wrap(err, "set relay log level")
+				return
+			}
+		}
+
+		// Increase resource limits
+		limiter := rcmgr.DefaultLimits
+		limiter.SystemBaseLimit.ConnsInbound = config.MaxConns
+		limiter.SystemBaseLimit.FD = config.MaxConns
+		limiter.TransientBaseLimit = limiter.SystemBaseLimit
+
+		mgr, err := rcmgr.NewResourceManager(rcmgr.NewFixedLimiter(limiter.Scale(1<<30, config.MaxConns))) // 1GB Memory
+		if err != nil {
+			p2pErr <- errors.Wrap(err, "new resource manager")
+		}
+
+		tcpNode, err := p2p.NewTCPNode(ctx, config.P2PConfig, key, p2p.NewOpenGater(), libp2p.ResourceManager(mgr))
+		if err != nil {
+			p2pErr <- errors.Wrap(err, "new tcp node")
+			return
+		}
+
+		log.SetLokiLabels(map[string]string{
+			"bootnode_peer": p2p.PeerName(tcpNode.ID()),
+		})
+
+		p2p.RegisterConnectionLogger(tcpNode, nil)
+
+		// Reservations are valid for 30min (github.com/libp2p/go-libp2p/p2p/protocol/circuitv2/relay/constraints.go:14)
+		relayResources := relay.DefaultResources()
+		relayResources.Limit.Data = 32 * (1 << 20) // 32MB
+		relayResources.MaxReservationsPerPeer = config.MaxResPerPeer
+		relayResources.MaxReservationsPerIP = config.MaxResPerPeer
+		relayResources.MaxReservations = config.MaxConns
+		relayResources.MaxCircuits = config.MaxResPerPeer
+
+		relayService, err := relay.New(tcpNode, relay.WithResources(relayResources))
+		if err != nil {
+			p2pErr <- err
+			return
+		}
+
+		logP2P = func() {
+			peers := make(map[peer.ID]bool)
+			conns := tcpNode.Network().Conns()
+			for _, conn := range conns {
+				peers[conn.RemotePeer()] = true
+			}
+			log.Info(ctx, "Libp2p TCP open connections",
+				z.Int("total", len(conns)),
+				z.Int("peers", len(peers)),
+			)
+		}
+
+		log.Info(ctx, "Libp2p TCP relay started",
+			z.Str("peer_name", p2p.PeerName(tcpNode.ID())),
+			z.Any("p2p_tcp_addr", config.P2PConfig.TCPAddrs),
+		)
+
+		<-ctx.Done()
+		_ = tcpNode.Close()
```

Comment:
why do we not handle error here?

Reply:
this is existing logic, I'm just moving it.